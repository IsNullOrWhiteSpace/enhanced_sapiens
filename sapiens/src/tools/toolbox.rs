
use std::collections::HashMap;
use std::fmt::Debug;
use std::sync::Arc;

use serde::{Deserialize, Serialize};
use tokio::sync::RwLock;
use tracing::{debug, info};

use crate::tools;
use crate::tools::invocation::InvocationError;
use crate::tools::{
    AdvancedTool, TerminalTool, TerminationMessage, Tool, ToolDescription, ToolUseError,
};