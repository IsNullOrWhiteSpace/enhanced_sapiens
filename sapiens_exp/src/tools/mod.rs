
//! Bake the acceptance test for the task in the tools
//! Scenario -> (Recipe, Config)
//!
//! Recipe -> (Task, Acceptance criteria)
//!
//! Acceptance criteria -> Tools
//!
//! run_until_completion: (Task, Config, Tools) -> (Trace, Stats, Acceptance
//! results)
//!
//! evaluate::Trial::analyze: (Trace, Stats, Acceptance results) -> Analysis
//!
//! evaluate::Trial::build: (Trace, State, Acceptance results, Task, Analysis)
//! -> Trial

use std::marker::PhantomData;
use std::ops::DerefMut;
use std::sync::Arc;

use sapiens::tools::{Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError};
use serde::{Deserialize, Serialize};
use tokio::sync::Mutex;

/// Scenario 0: Preparing a cereal bowl
pub mod scenario_0;

mod scenario_1;

/// The state of a scenario
pub trait State: Send + Sync {
    /// Reset the state to its initial state
    fn reset(&mut self);
