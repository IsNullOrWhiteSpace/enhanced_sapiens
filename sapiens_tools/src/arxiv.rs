use std::fmt::Display;

use arxiv;
use arxiv::{Arxiv, ArxivQuery};
use sapiens::tools::{Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError};
use sapiens_derive::{Describe, ProtoToolDescribe, ProtoToolInvoke};
use serde::{Deserialize, Serialize};

/// A Tool to query arXiv.
///
/// arXiv is a free distribution service and an open-access archive for
/// millions scholarly articles in the fields of physics, mathematics, computer
/// science, quantitative biology, quantitative finance, statistics, electrical
/// engineering and systems science, and economics. Materials on this site are
/// not peer-reviewed by arXiv.
#[derive(Debug, ProtoToolInvoke, ProtoToolDescribe)]
#[tool(name = "Arxiv", input = "ArxivToolInput", output = "ArxivToolOutput")]
pub struct ArxivTool {}

/// Sort order
#[derive(Debug, Deserialize, Serialize, Default, Clone)]
pub enum SortOrder {
    /// Ascending
    #[serde(rename = "ascending")]
    Ascending,
    /// Descending
    #[serde(rename = "descending")]
    #[default]
    Descending,
}

impl Display for SortOrder {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SortOrder::Ascending => write!(f, "ascending"),
            SortOrder::Descending => write!(f, "descending"),
        }
    }
}

/// Sort by
#[derive(Debug, Deserialize, Serialize, Default, Clone)]
pub enum SortBy {
    /// Relevance
    #[serde(rename = "relevance")]
    #[default]
    Relevance,
    /// Last updated date
    #[serde(rename = "lastUpdatedDate")]
    LastUpdatedDate,
    /// Submitted date
    #[serde(rename = "submittedDate")]
    SubmittedDate,
}

impl Display for SortBy {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        match self {
            SortBy::Relevance => write!(f, "relevance"),
            SortBy::LastUpdatedDate => write!(f, "lastUpdatedDate"),
            SortBy::SubmittedDate => write!(f, "submittedDate"),
        }
    }
}

/// [`ArxivTool`] input
///
/// Arxiv API documentation query specification
#[derive(Debug, Deserialize, Serialize, Describe)]
pub struct ArxivToolInput {
    /// search_query: Search query - see https://info.arxiv.org/help/api/user-manual.html
    /// for details. E.g. `cs.AI` or `cat:cs.AI` or `au:John Smith`
    /// The fields that can be searched are: `ti` (title), `au` (author), `abs`
    /// (abstract), `co` (comment), `jr` (journal reference), `cat` (subject
    /// category), `rn` (report number), `id` (id (use id_list instead)),
    /// `all` (all of the above). Operators: `AND`, `OR`, `ANDNOT`.
    /// You cannot search on publication or last update date.
    pub search_query: String,

    /// id_list: Comma-separated list of arXiv IDs to return
    pub id_list: Option<String>,

    /// start: Result offset for pagination
    pub start: Option<i32>,

    /// max_results: Maximum number of results to return in a single response.
    /// Default is 10. Maximum allowed value is 100.
    pub max_results: Option<i32>,

    /// Sort by. Can be either `relevance`, `lastUpdatedDate` or
    /// `submittedDate`. Default is `relevance`.
    pub sort_by: Option<SortBy>,

    /// Sort order. Can be either `ascending` or `descending`.
    /// Default is `descending`.
    pub sort_order: Option<SortOrder>,

    /// True to gather PDF url - default is false
    pub show_pdf_url: Option<bool>,

    /// True to gather authors - default is false
    pub show_authors: Option<bool>,

    /// True to gather comments - default is false
    pub show_comments: Option<bool>,

    /// True to gather summary - default is false
    pub show_summary: Option<bool>,
}

impl From<&ArxivToolInput> for ArxivQuery {
    fn from(input: &ArxivToolInput) -> Self {
        ArxivQuery {
            base_url: "https://export.arxiv.org/api/query?".to_string(),
            search_query: input.search_query.clone(),
            id_list: input.id_list.clone().unwrap_or_default(),
            start: input.start,
            max_results: input.max_results,
            sort_by: input.sort_by.clone().unwrap_or_default().to_string(),
            sort_order: input.sort_order.clone().unwrap_or_default().to_string(),
        }
    }
}

/// [`ArxivTool`] output
#[derive(Debug, Deserialize, Serialize, Describe)]
pub struct ArxivToolOutput {
    // FUTURE(ssoudan) proc_macro_derive to generate this
    /// query result. `ArxivResult` is an object containing the following
    /// fields:
    /// - `id`: <str> arXiv ID
    /// - `updated`: <str> last updated date
    /// - `published`: <str> published date
    /// - `title`: <str> title
    /// - `summary`: <Optional[str]> summary - omitted unless `show_summary` is
    ///   true - can be quite long
    /// - `authors`: <list[str]> authors - omitted unless `show_authors` is true
    /// - `pdf_url`: <Optional[str]> PDF URL - omitted unless `show_pdf_url` is
    ///   true
    /// - `comments`: <Optional[str]> Comments - omitted unless `show_comments`
    ///   is true
    result: Vec<ArxivResult>,
}

/// Arxiv result
#[derive(Debug, Deserialize, Serialize, Describe)]
pub struct ArxivResult {
    /// arXiv ID
    pub id: String,
    /// last updated date
    pub updated: String,
    /// published date
    pub published: String,
    /// title
    pub title: String,
    /// summary - only if `show_summary` is true - can be quite long
    #[serde(skip_serializing_if = "Option::is_none")]
    pub summary: Option<String>,
    /// authors - only if `show_authors` is true
    #[serde(skip_serializing_if = "Vec::is_empty")]
    pub authors: Vec<String>,
    /// PDF URL - only if `show_pdf_url` is true
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pdf_url: Option<String>,
    /// Comments - only if `show_comments` is true
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comment: Option<String>,
}

impl From<Arxiv> for ArxivResult {
    fn from(arxiv: Arxiv) -> Self {
        ArxivResult {
            id: arxiv.id,
            updated: arxiv.updated,
            published: arxiv.published,
            title: arxiv.title,
            summary: Some(arxiv.summary),
            authors: arxiv.authors,
            pdf_url: Some(arxiv.pdf_url),
            comment: arxiv.comment,
        }
    }
}

impl ArxivTool {
    /// Create a new [`ArxivTool`]
    pub async fn new() -> ArxivTool {
        ArxivTool {}
    }

    #[tracing::instrument(skip(self))]
    async fn invoke_typed(&self, input: &ArxivToolInput) -> Result<ArxivToolOutput, ToolUseError> {
        let query = ArxivQuery::from(input);

        if query.max_results.unwrap_or(0) > 100 {
            return Err(ToolUseError::InvocationFailed(
                "max_results cannot be greater than 100".to_string(),
            ));
        }
