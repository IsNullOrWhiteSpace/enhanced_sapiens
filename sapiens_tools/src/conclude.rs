
use std::fmt::Debug;

use sapiens::tools::{
    Describe, ProtoToolDescribe, ProtoToolInvoke, TerminalTool, TerminationMessage,
    ToolDescription, ToolUseError,
};
use sapiens_derive::{Describe, ProtoToolDescribe, ProtoToolInvoke};
use serde::{Deserialize, Serialize};
use tokio::sync::Mutex;

/// A tool to conclude a task.
/// You have to use this to once you have the answer to the task with your
/// conclusion.
#[derive(Default, ProtoToolDescribe, ProtoToolInvoke)]
#[tool(
    name = "Conclude",
    input = "ConcludeToolInput",
    output = "ConcludeToolOutput"
)]
pub struct ConcludeTool {
    done: Mutex<Option<ConcludeToolInput>>,
}

impl Debug for ConcludeTool {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.debug_struct("ConcludeTool").finish()
    }
}

#[async_trait::async_trait]
impl TerminalTool for ConcludeTool {
    async fn is_done(&self) -> bool {
        // lock
        let done = self.done.lock().await;
        done.is_some()
    }

    async fn take_done(&self) -> Option<TerminationMessage> {
        // lock
        {
            let mut done = self.done.lock().await;
            done.take().map(|input| input.into())
        }
    }
}

/// A tool that is called to wrap the task.
#[derive(Debug, Clone, Serialize, Deserialize, Describe)]
pub struct ConcludeToolInput {
    /// The final answer for this task. Plain text ONLY. No string interpolation
    /// supported. MANDATORY. Call directly from `SandboxPython` Tool for long
    /// answers.
    pub conclusion: String,
    /// The original question that was asked to the user. No string
    /// interpolation supported, only plain text. MANDATORY.
    pub original_question: String,
}

impl From<ConcludeToolInput> for TerminationMessage {
    fn from(input: ConcludeToolInput) -> Self {
        Self {
            conclusion: input.conclusion,
            original_question: input.original_question,
        }
    }
}

/// ConcludeToolOutput - empty
#[derive(Serialize, Deserialize, Describe)]
pub struct ConcludeToolOutput {}

impl ConcludeTool {
    #[tracing::instrument(skip(self))]
    async fn invoke_typed(
        &self,
        input: &ConcludeToolInput,
    ) -> Result<ConcludeToolOutput, ToolUseError> {
        // lock
        {
            let mut done = self.done.lock().await;

            if done.is_some() {
                return Err(ToolUseError::InvocationFailed(
                    "This task is already done.".to_string(),
                ));
            }

            // set done
            *done = Some(input.clone());
        }

        Ok(ConcludeToolOutput {})
    }
}