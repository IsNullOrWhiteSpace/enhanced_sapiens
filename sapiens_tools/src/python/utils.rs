
use pyo3::types::{IntoPyDict, PyDict, PyFloat, PyList, PyTuple};
use pyo3::{PyObject, Python, ToPyObject};
use sapiens::tools::{FieldFormat, ToolDescription};
use serde::{Deserialize, Serialize};
use serde_yaml::Value;

#[derive(thiserror::Error, Debug)]
pub(crate) enum PyConversionError {
    #[error("Invalid conversion: {error}")]
    InvalidConversion { error: String },
    #[error("dict key not serializable: {typename}")]
    DictKeyNotSerializable { typename: String },
    #[error("Invalid cast: {typename}")]
    InvalidCast { typename: String },
}

// inspired from https://github.com/mozilla-services/python-canonicaljson-rs/blob/62599b246055a1c8a78e5777acdfe0fd594be3d8/src/lib.rs#L87-L167
#[allow(clippy::redundant_closure_call)]
pub(crate) fn to_yaml(py: Python, obj: &PyObject) -> Result<Value, PyConversionError> {
    macro_rules! return_cast {
        ($t:ty, $f:expr) => {
            if let Ok(val) = obj.downcast::<$t>(py) {
                return $f(val);
            }
        };
    }

    macro_rules! return_to_value {
        ($t:ty) => {
            if let Ok(val) = obj.extract::<$t>(py) {
                return serde_yaml::to_value(val).map_err(|error| {
                    PyConversionError::InvalidConversion {
                        error: format!("{}", error),
                    }
                });
            }
        };
    }

    if obj.is_none(py) {
        return Ok(Value::Null);
    }

    return_to_value!(String);
    return_to_value!(bool);
    return_to_value!(u64);
    return_to_value!(i64);

    return_cast!(PyDict, |x: &PyDict| {
        let mut map = serde_yaml::Mapping::new();
        for (key_obj, value) in x.iter() {
            let key = if key_obj.is_none() {
                Ok("null".to_string())
            } else if let Ok(val) = key_obj.extract::<bool>() {
                Ok(if val {
                    "true".to_string()
                } else {
                    "false".to_string()
                })
            } else if let Ok(val) = key_obj.str() {
                Ok(val.to_string())
            } else {
                Err(PyConversionError::DictKeyNotSerializable {
                    typename: key_obj
                        .to_object(py)
                        .as_ref(py)
                        .get_type()
                        .name()
                        .map(|x| x.to_string())
                        .unwrap_or_else(|_| "unknown".to_string()),
                })
            };
            map.insert(Value::String(key?), to_yaml(py, &value.to_object(py))?);
        }
        Ok(Value::Mapping(map))
    });

    return_cast!(PyList, |x: &PyList| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyTuple, |x: &PyTuple| {
        let v = x
            .iter()
            .map(|x| to_yaml(py, &x.to_object(py)))
            .collect::<Result<Vec<_>, _>>()?;
        Ok(Value::Sequence(v))
    });

    return_cast!(PyFloat, |x: &PyFloat| {
        Ok(Value::Number(serde_yaml::Number::from(x.value())))
    });

    // At this point we can't cast it, set up the error object
    Err(PyConversionError::InvalidCast {
        typename: obj
            .as_ref(py)
            .get_type()
            .name()
            .map(|x| x.to_string())
            .unwrap_or_else(|_| "unknown".to_string()),
    })
}

pub(crate) fn value_to_object(val: Value, py: Python<'_>) -> PyObject {
    match val {
        Value::Null => py.None(),
        Value::Bool(x) => x.to_object(py),
        Value::Number(x) => {
            let oi64 = x.as_i64().map(|i| i.to_object(py));
            let ou64 = x.as_u64().map(|i| i.to_object(py));
            let of64 = x.as_f64().map(|i| i.to_object(py));
            oi64.or(ou64).or(of64).expect("number too large")
        }
        Value::String(x) => x.to_object(py),
        Value::Sequence(x) => {
            let inner: Vec<_> = x.into_iter().map(|x| value_to_object(x, py)).collect();
            inner.to_object(py)
        }
        Value::Mapping(x) => {
            let iter = x
                .into_iter()
                .map(|(k, v)| (value_to_object(k, py), value_to_object(v, py)));
            IntoPyDict::into_py_dict(iter, py).into()
        }
        Value::Tagged(_) => panic!("tagged values are not supported"),
    }
}

/// Format of a field in a tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct SimpleFormat {
    /// Name of the field
    pub name: String,
    /// Type of the field
    pub r#type: String,
    /// Whether the field is optional
    pub optional: bool,
    /// Description of the field
    pub description: String,
}

impl From<FieldFormat> for SimpleFormat {
    fn from(part: FieldFormat) -> Self {
        SimpleFormat {
            name: part.name,
            r#type: part.r#type,
            optional: part.optional,
            description: part.description,
        }
    }
}

impl ToPyObject for SimpleFormat {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.to_object(py)).unwrap();
        dict.set_item("type", self.r#type.to_object(py)).unwrap();
        dict.set_item("optional", self.optional.to_object(py))
            .unwrap();
        dict.set_item("description", self.description.to_object(py))
            .unwrap();
        dict.into()
    }
}

/// A simplified version of the tool description
#[derive(Debug, Clone, Serialize, Deserialize)]
pub(crate) struct SimpleToolDescription {
    /// Name of the tool
    pub name: String,
    /// Description of the tool
    pub description: String,
    /// Input format of the tool
    pub parameters: Vec<SimpleFormat>,
    /// Output format of the tool
    pub responses_content: Vec<SimpleFormat>,
}

impl From<ToolDescription> for SimpleToolDescription {
    fn from(desc: ToolDescription) -> Self {
        let parameters = desc
            .parameters
            .fields
            .into_iter()
            .map(|x| x.into())
            .collect();
        let responses_content = desc
            .responses_content
            .fields
            .into_iter()
            .map(|x| x.into())
            .collect();

        SimpleToolDescription {
            name: desc.name,
            description: desc.description,
            parameters,
            responses_content,
        }
    }
}

impl ToPyObject for SimpleToolDescription {
    fn to_object(&self, py: Python<'_>) -> PyObject {
        let dict = PyDict::new(py);
        dict.set_item("name", self.name.clone()).unwrap();
        dict.set_item("description", self.description.clone())
            .unwrap();
        dict.set_item("parameters", self.parameters.clone())
            .unwrap();
        dict.set_item("responses_content", self.responses_content.clone())
            .unwrap();
        dict.into()
    }
}