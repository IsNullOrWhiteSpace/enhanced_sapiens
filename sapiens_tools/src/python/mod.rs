
use std::cmp::Ordering;
use std::collections::HashMap;

use convert_case::{Case, Casing};
use pyo3::indoc::{formatdoc, indoc};
use pyo3::prelude::*;
use pyo3::types::{IntoPyDict, PyDict};
use sapiens::tools::toolbox::{invoke_simple_from_toolbox, Toolbox};
use sapiens::tools::{
    AdvancedTool, Describe, ProtoToolDescribe, ProtoToolInvoke, ToolDescription, ToolUseError,
};
use sapiens_derive::{Describe, ProtoToolDescribe};
use serde::{Deserialize, Serialize};
use serde_yaml::Value;
use tracing::trace;

/// Conversion tools
pub(crate) mod utils;

use crate::python::utils::SimpleToolDescription;

const MAX_OUTPUT_SIZE: usize = 512;

// FUTURE(ssoudan) install pySWIP

/// A tool that runs sandboxed Python code. Use this to transform data.
///
/// - To use another Tool with parameters `input_field_1` and `input_field_2`
///   and result fields `output_field_1` and `output_field_2` use:
/// ```python
/// result = tools.ToolName(input_field_1=..., input_field_2=...)
/// print(result['output_field_1'])
/// print(result['output_field_2'])
/// ```
/// - Only stdout and stderr are captured and made available (limited to 512B
///   total). If the output is larger, use `tools.Conclude` directly from the
///   code.
/// - List available tools with `tools.list()`. And returns a list of
///   `{'name':.., 'description':.., 'parameters':.., 'responses_content':..,
///   }`.
/// - `open`|`exec` are forbidden.
/// - Limited libraries available: urllib3, requests, sympy, numpy,
/// BeautifulSoup4, feedparser, arxiv.
/// - No PIP.
#[derive(Debug, Default, ProtoToolDescribe)]
#[tool(
    name = "SandboxedPython",
    input = "PythonToolInput",
    output = "PythonToolOutput"
)]
pub struct PythonTool {}

/// The input of the Python tool
#[derive(Debug, Serialize, Deserialize, Describe)]
pub struct PythonToolInput {
    /// The Python code to run. MANDATORY
    pub code: String,
}

/// The output of the Python tool
#[derive(Serialize, Deserialize, Describe)]
pub struct PythonToolOutput {
    /// The stdout output of the Python code.
    pub stdout: String,
    /// The stderr output of the Python code.
    pub stderr: String,
}

#[pyclass]
#[derive(Default)]
struct Logging {
    output: String,
}

#[pymethods]
impl Logging {
    fn write(&mut self, data: &str) {
        self.output.push_str(data);
    }
}

#[pyclass(unsendable)]
struct ToolsWrapper {
    toolbox: Toolbox,
    tool_list: Vec<SimpleToolDescription>,
}

impl ToolsWrapper {
    async fn new(toolbox: Toolbox) -> Self {
        let tools = toolbox.describe().await;
        let tool_list = tools
            .into_values()
            .map(SimpleToolDescription::from)
            .collect::<Vec<_>>();

        ToolsWrapper { toolbox, tool_list }
    }
}

#[pymethods]
impl ToolsWrapper {
    // list all tools
    #[pyo3(signature = ())]
    fn list(&self, py: Python<'_>) -> PyResult<PyObject> {
        let tools = self.tool_list.to_object(py);
        Ok(tools)
    }

    // invoke a tool
    #[pyo3(signature = (tool_name, input))]
    fn invoke(
        &self,
        py: Python<'_>,
        tool_name: &str,
        input: Option<&PyDict>,
    ) -> PyResult<PyObject> {
        // convert PyDict to a serde_yaml::Value
        let input = if let Some(input) = input {
            let input: PyObject = input.into();

            utils::to_yaml(py, &input).map_err(|e| {
                pyo3::exceptions::PyException::new_err(format!("Invalid input: {}", e))
            })?
        } else {
            Value::default()
        };

        let (tx, mut rx) = tokio::sync::oneshot::channel::<Result<Value, ToolUseError>>();

        // release the GIL to allow the thread to run
        py.allow_threads(move || {