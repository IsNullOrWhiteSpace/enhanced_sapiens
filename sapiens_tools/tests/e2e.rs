
//! Tests
#[cfg(feature = "hue")]
mod hue_test {
    use indoc::indoc;
    use insta::{assert_display_snapshot, assert_yaml_snapshot};
    use pyo3::PyResult;
    use sapiens::tools::toolbox::{invoke_tool, InvokeResult, Toolbox};
    use sapiens_tools::conclude::ConcludeTool;
    use sapiens_tools::hue::room::fake::FakeRoomTool;
    use sapiens_tools::hue::status::fake::FakeStatusTool;
    use sapiens_tools::python::PythonTool;

    #[pyo3_asyncio::tokio::test]
    async fn test_python_hue() -> PyResult<()> {
        let data = indoc! {r#"
    # Action
    ```yaml
    tool_name: SandboxedPython
    parameters:
      code: |
        room_name = "Bedroom"
    
        import json
    
        # Get the list of room objects with names and Lights ID lists.
        rooms = tools.Room(room_filter=[]).get('rooms')
    
        # Find the relevant room and get only its lights IDs.
        room_lights = next(room for room in rooms if room_name in room.get('name'))
        room_lights_ids = room_lights.get('lights')
                        
        # Get the statuses of Lights in the Bedroom.
        Bedroom_lights_status = tools.LightStatus(light_filter=room_lights_ids).get('lights')
    
        # Conclude the task with the result.
        tools.Conclude(
            original_question="what is the status of the lights in the Bedroom?",
            conclusion=f"The light(s) status in {room_name} is {json.dumps(Bedroom_lights_status)}"
        )
    ```
    "#};

        let mut toolbox = Toolbox::default();
        toolbox.add_advanced_tool(PythonTool::default()).await;
        toolbox.add_terminal_tool(ConcludeTool::default()).await;
        toolbox.add_tool(FakeRoomTool::default()).await;
        toolbox.add_tool(FakeStatusTool::default()).await;

        let res = invoke_tool(toolbox.clone(), data).await;

        match res {
            InvokeResult::Success {
                tool_name, result, ..
            } => {
                assert_eq!(tool_name, "SandboxedPython");
                assert_display_snapshot!(result);
            }
            _ => panic!("Unexpected response: {:?}", res),
        }

        // collect the conclusion
        let termination_messages = toolbox.termination_messages().await;

        assert_yaml_snapshot!(termination_messages);

        Ok(())
    }
}

#[cfg(feature = "arxiv")]
mod arxiv {
    use indoc::indoc;
    use insta::{assert_display_snapshot, assert_yaml_snapshot};
    use pyo3::PyResult;
    use sapiens::tools::toolbox::{invoke_tool, InvokeResult, Toolbox};
    use sapiens_tools::arxiv::ArxivTool;
    use sapiens_tools::conclude::ConcludeTool;
    use sapiens_tools::python::PythonTool;

    #[pyo3_asyncio::tokio::test]
    async fn test_python_arxiv() -> PyResult<()> {
        let mut toolbox = Toolbox::default();
        toolbox.add_tool(ArxivTool::new().await).await;
        toolbox.add_advanced_tool(PythonTool::default()).await;

        let data = indoc! {r#"```yaml
   tool_name: SandboxedPython
   parameters:
     code: |           
           import datetime               
           search_query = f'cat:cs.AI'
           input = {
               'search_query': search_query,
               'start': 0,
               'max_results': 10,
               'sort_by': 'lastUpdatedDate',
               'sort_order': 'ascending'
           }
           arxiv_output = tools.Arxiv(**input)
           papers = arxiv_output['result']               
           conclusion = f"Yes, there are {len(papers)} published papers of AI category on ArXiv this week." if papers else "No, there are no published papers of AI category on ArXiv this week."
           print(conclusion)
   ```
"#};

        let res = invoke_tool(toolbox, data).await;

        match res {
            InvokeResult::Success {
                tool_name, result, ..
            } => {
                assert_eq!(tool_name, "SandboxedPython");
                assert_display_snapshot!(result);
            }
            _ => panic!("Unexpected response: {:?}", res),
        }

        Ok(())
    }

    #[pyo3_asyncio::tokio::test]
    async fn test_python_arxiv_2() -> PyResult<()> {
        let mut toolbox = Toolbox::default();
        toolbox.add_tool(ArxivTool::new().await).await;
        toolbox.add_advanced_tool(PythonTool::default()).await;

        let data = indoc! {r#"```yaml
   tool_name: SandboxedPython
   parameters:
     code: |           
           import datetime               
           search_query = f'cat:cs.AI'               
           arxiv_output = tools.Arxiv(search_query=search_query, start=0, max_results=10, sort_by='lastUpdatedDate', sort_order='ascending')               
           papers = arxiv_output['result']               
           conclusion = f"Yes, there are {len(papers)} published papers of AI category on ArXiv this week." if papers else "No, there are no published papers of AI category on ArXiv this week."
           print(conclusion)
   ```
"#};

        let res = invoke_tool(toolbox, data).await;

        match res {
            InvokeResult::Success {
                tool_name, result, ..
            } => {
                assert_eq!(tool_name, "SandboxedPython");
                assert_display_snapshot!(result);
            }
            _ => panic!("Unexpected response: {:?}", res),
        }

        Ok(())
    }

    #[pyo3_asyncio::tokio::test]
    async fn test_python_arxiv_3() -> PyResult<()> {
        let mut toolbox = Toolbox::default();
        toolbox.add_tool(ArxivTool::new().await).await;
        toolbox.add_advanced_tool(PythonTool::default()).await;

        let data = indoc! {r#"```yaml
   tool_name: SandboxedPython
   parameters:
     code: |
        latest_papers = tools.Arxiv(
          search_query="cat:cs.DB",
          max_results=4,
          show_summary=True,
          show_pdf_url=True
          )["result"]           
        print(latest_papers)
   ```
"#};

        let res = invoke_tool(toolbox, data).await;

        match res {
            InvokeResult::Success {
                tool_name, result, ..
            } => {
                assert_eq!(tool_name, "SandboxedPython");
                assert_display_snapshot!(result);
            }
            _ => panic!("Unexpected response: {:?}", res),
        }

        Ok(())
    }

    #[pyo3_asyncio::tokio::test]
    async fn test_python_arxiv_4() -> PyResult<()> {
        let mut toolbox = Toolbox::default();
        toolbox.add_tool(ArxivTool::new().await).await;
        toolbox.add_terminal_tool(ConcludeTool::default()).await;
        toolbox.add_advanced_tool(PythonTool::default()).await;

        let data = indoc! {r#"```yaml
    tool_name: SandboxedPython
    parameters:
      code: |
        latest_papers = tools.Arxiv(
          search_query="cat:cs.DB",
          max_results=4,
          show_summary=True,
          show_pdf_url=True,
          sort_order="ascending",
          )["result"]
    
        summaries = []
        for paper in latest_papers:
            summary = paper["summary"]
            summaries.append(summary)
    
        task_conclusion = "\n\n\n".join([f"Title: {paper['title']}\nSummary: {summary}\nPDF Url: {paper['pdf_url']}"
                                        for paper, summary in zip(latest_papers, summaries)])
    
        tools.Conclude(
                original_question="What are the 4 latest papers on database published on arXiv? What are they about?",
                conclusion=task_conclusion)
   ```
"#};

        let res = invoke_tool(toolbox.clone(), data).await;

        match res {
            InvokeResult::Success {
                tool_name, result, ..
            } => {
                assert_eq!(tool_name, "SandboxedPython");
                assert_display_snapshot!(result);
            }
            _ => panic!("Unexpected response: {:?}", res),
        }

        let termination_messages = toolbox.termination_messages().await;
        let done = !termination_messages.is_empty();
        assert!(done);

        assert_yaml_snapshot!(termination_messages);

        Ok(())
    }
}

mod python;

#[cfg(all(feature = "search", not(feature = "disable-test-dependabot")))]
mod search {
    use indoc::indoc;
    use insta::assert_display_snapshot;
    use pyo3::PyResult;
    use sapiens::tools::toolbox::{invoke_tool, InvokeResult, Toolbox};
    use sapiens_tools::python::PythonTool;
    use sapiens_tools::search::SearchTool;

    #[pyo3_asyncio::tokio::test]
    async fn test_python_search() -> PyResult<()> {
        let mut toolbox = Toolbox::default();
        toolbox.add_tool(SearchTool::default()).await;
        toolbox.add_advanced_tool(PythonTool::default()).await;

        let data = indoc! {r#"```yaml
   tool_name: SandboxedPython
   parameters:
     code: |                
         input = {"q": "rust" }     
         output = tools.Search(**input)                                     
         print(len(output['results']))
   ```
"#};

        let res = invoke_tool(toolbox, data).await;

        match res {
            InvokeResult::Success {
                tool_name, result, ..
            } => {
                assert_eq!(tool_name, "SandboxedPython");
                assert_display_snapshot!(result);
            }
            _ => panic!("Unexpected response: {:?}", res),
        }

        Ok(())
    }
}

#[pyo3_asyncio::tokio::main(flavor = "multi_thread")]
async fn main() -> pyo3::PyResult<()> {
    let _ = dotenvy::dotenv();

    pyo3_asyncio::testing::main().await
}