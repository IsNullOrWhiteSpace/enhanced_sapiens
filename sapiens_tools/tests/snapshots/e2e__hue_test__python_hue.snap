---
source: sapiens_tools/tests/e2e.rs
expression: output.result
---
stdout: ''
stderr: ''

